homeassistant/components/synology_dsm/config_flow.py

```diff
@@ -131,12 +143,69 @@ class SynologyDSMFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
             CONF_PASSWORD: password,
         }
         if user_input.get(CONF_DISKS):
-            config_data.update({CONF_DISKS: user_input[CONF_DISKS]})
+            config_data[CONF_DISKS] = user_input[CONF_DISKS]
         if user_input.get(CONF_VOLUMES):
-            config_data.update({CONF_VOLUMES: user_input[CONF_VOLUMES]})
+            config_data[CONF_VOLUMES] = user_input[CONF_VOLUMES]
+
+        return self.async_create_entry(title=host, data=config_data)
+
+    def _login_and_fetch_syno_info(self, api):
+        """Login to the NAS and fetch basic data."""
+        if not api.login():
+            raise InvalidAuth
+
+        # These do i/o
+        information = api.information
+        utilisation = api.utilisation
+        storage = api.storage
+
+        if (
+            information.serial is None
+            or utilisation.cpu_user_load is None
+            or storage.disks_ids is None
+            or storage.volumes_ids is None
+        ):
+            raise InvalidData
+
+        return information.serial
 
-        return self.async_create_entry(title=host, data=config_data,)
+    async def async_step_ssdp(self, discovery_info):
+        """Handle a discovered synology_dsm."""
+        parsed_url = urlparse(discovery_info[ssdp.ATTR_SSDP_LOCATION])
+        friendly_name = (
+            discovery_info[ssdp.ATTR_UPNP_FRIENDLY_NAME].split("(", 1)[0].strip()
+        )
+
+        if self._host_already_configured(parsed_url.hostname):
+            return self.async_abort(reason="already_configured")
+
+        self.discovered_conf = {
+            CONF_NAME: friendly_name,
+            CONF_HOST: parsed_url.hostname,
+        }
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        self.context["title_placeholders"] = self.discovered_conf
+        return await self.async_step_user()
 
     async def async_step_import(self, user_input=None):
         """Import a config entry."""
         return await self.async_step_user(user_input)
+
+    async def async_step_link(self, user_input=None):
+        """Link a config entry from discovery."""
+        return await self.async_step_user(user_input)
+
+    def _host_already_configured(self, hostname):
+        """See if we already have a host matching user input configured."""
+        existing_hosts = {
+            entry.data[CONF_HOST] for entry in self._async_current_entries()
+        }
+        return hostname in existing_hosts
```

Comment:
The host is an IP from the local network here, right ? (first time handling discovery)

Reply:
Yes. It is actually an ip address.  CONF_HOST will take either an ip or host

Reply:
OK nice then.

I thought of a case when using one DNS with multiple ports to access multiple NAS.